### ISSUE TITLE
Health checked record support

### ISSUE THREAD

**Issue by dolvany** [github/octodns#49]
First of all, thank you for this beautiful piece of software. I have set up a traffic director on dyn and added `traffic_directors_enabled: true` to my dyn provider config, but no traffic director records appear in a dump. Any guidance getting the traffic director feature to work is much appreciated.

**Comment by ross** [comment]
Hi @dolvany. You might try adding a `--debug` to the `octodns-dump` command line to see if the verbose logging sheds any light on what's happening. There won't be a lot of information, but you should see something like the following for each TD it works through and they'll generally come through pretty slowly as there's several API calls involved with each. 

```
...
017-06-01T04:54:42  [140735274140416] DEBUG DynProvider[dyn] _populate_traffic_directors: zone=github.com.
2017-06-01T04:54:50  [140735274140416] DEBUG Record __init__: zone.name=github.com., type=    ARecord, name=
...
```

You may want to crack open `DynProvider` and add some more debug-level logging to the TD handling paths to help track down what's happening. Feel free to PR that if it seems useful.

After that my guess would be there are differences between the way octoDNS structures TD records and what you may have set up. Due to trying to get the same behavior across multiple providers octoDNS is pretty opinionated about how TD records are structured. It may be that something about the way your stuff is configured doesn't match with what it's looking for. If this seems likely I'll probably need to know more details about your setup to be able to help. If you're not comfortable sharing real records/zone info here maybe it'd made sense to create a dummy/test zone with a similar structure that you could share. Failing that we can probably connect over email if you'd be OK with sharing details there.

**Comment by ross** [comment]
Took a look at the debug log & a snapshot of the config out of band in email and coming back here with the result.

In the specific problematic case TD is being used for health checking purposes. That's not a use-case currently supported, though is an interesting one that it'd be nice to have. A lot of the necessary bits already exist in the code as health checks are part of `geo` support, but they're plumbed differently. 

Internally our IP-redundancy [happens at the network level](https://githubengineering.com/introducing-glb/), but our previous load balancer arch did make use of this sort of config. 

The following is a rough sketch of what I think a health-check only TD setup would probably look like. The interesting bit is that there's a `checked` pool that's normally in use that serves all the healthy IPs. If one of them were to fail it would just serve the other. If both of them were to fail it would fall back to serving both records again as a safety valve. The benefits to that setup is that if something goes really wrong it'll keep trying anyway rather than serving no records. This would be useful in the case that the health check endpoint was busted, but the overlying services weren't. It would also get rid of the recovery time lag when the service becomes healthy after problems since they'd be sent requests anyway so they'd immediately start answering them while the DNS health checks/status catches up.

I haven't looked at Route53 (and the other providers with health checking) so it's possible this would change if/when a unified solution was designed and implemented.

![screen shot 2017-06-01 at 7 11 24 am](https://cloud.githubusercontent.com/assets/12789/26683837/02e37dc8-469a-11e7-93b2-fba3891a39d0.png)

For reference here's what a (mocked up) GeoDNS setup built by octoDNS looks like, there should really be a doc about this, but until someone gets time to put that together this should be useful:

![screen shot 2017-06-01 at 7 15 14 am](https://cloud.githubusercontent.com/assets/12789/26683871/24a2efc0-469a-11e7-999c-310cb7ce6c63.png)

The only interesting bit that isn't shown above is how `NA-US-CA` falls back to `NA-US` before going to `default`.

![screen shot 2017-06-01 at 7 16 01 am](https://cloud.githubusercontent.com/assets/12789/26683923/53b23f14-469a-11e7-9d1f-d7cd0691c0ce.png)

TL;DR is that straight health checks aren't currently supported, but it'd be nice to have them. Not an immediate need I have so no clue on what the timeline would look like for them. Unfortunately it's not really the sort of thing that can be knocked out in nights & weekends as the APIs for working with more complex setups are non-trivial.

**Comment by ross** [comment]
In theory you may be able work around this for Dyn by doing something like:

```
'':
  geo:
    AF: 
    - 1.2.3.4
    - 1.2.3.5
    AN:
    - 1.2.3.4
    - 1.2.3.5
    AS: 
    - 1.2.3.4
    - 1.2.3.5
    EU: 
    - 1.2.3.4
    - 1.2.3.5
    NA: 
    - 1.2.3.4
    - 1.2.3.5
    OC:
    - 1.2.3.4
    - 1.2.3.5
    SA: 
    - 1.2.3.4
    - 1.2.3.5
  ttl: 60
  values:
  - 1.2.3.4
  - 1.2.3.5
```

That basically sets up geo-ip records for all of the continents which "should" cover everything. This will work due to the way that dyn health checks check each IP individually. It **won't** work like that with Route53 as it health checks each record rather than IP so it's really kind of a fluke that it works for Dyn. To get the same behavior with Route53 it looks like you'd have to use a completely different type of record.

**Comment by dolvany** [comment]
@ross, thanks for the pointers on the geo functionality. One followup q: Is it possible to mark a record for noop? Like if I want to prevent replication of a record from one provider to another.

**Comment by ross** [comment]
> Is it possible to mark a record for noop? Like if I want to prevent replication of a record from one provider to another.

Not currently. Would be easy to add something like `ignore: true` to the config provider. It'd take a little thinking to figure out how to plumb that through and to get the changes process to do the right things with them. Feel free to file an issue specifically requesting that. It seems like a reasonable thing to support at some point.

**Comment by ross** [comment]
Pass at the `ignored` bit of things over in https://github.com/github/octodns/pull/56.

**Comment by ross** [comment]
This should be possible now with the new `dynamic` record type.
